compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/scene/ComposeScene.kt

```diff
@@ -0,0 +1,272 @@
+import androidx.compose.ui.input.pointer.PointerEventType
+import androidx.compose.ui.input.pointer.PointerKeyboardModifiers
+import androidx.compose.ui.input.pointer.PointerType
+import androidx.compose.ui.node.LayoutNode
+import androidx.compose.ui.platform.LocalLayoutDirection
+import androidx.compose.ui.platform.setContent
+import androidx.compose.ui.semantics.SemanticsNode
+import androidx.compose.ui.semantics.SemanticsOwner
+import androidx.compose.ui.unit.Constraints
+import androidx.compose.ui.unit.Density
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.IntRect
+import androidx.compose.ui.unit.IntSize
+import androidx.compose.ui.unit.LayoutDirection
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.window.Dialog
+import androidx.compose.ui.window.Popup
+import org.jetbrains.skiko.currentNanoTime
+
+/**
+ * Represents a static [CompositionLocal] key for a [ComposeScene] in Jetpack Compose.
+ *
+ * @see ComposeScene
+ */
+@OptIn(InternalComposeUiApi::class)
+internal val LocalComposeScene = staticCompositionLocalOf<ComposeScene?> { null }
+
+/**
+ * The local [ComposeScene] is typically not-null. This extension can be used in these cases.
+ */
+@OptIn(InternalComposeUiApi::class)
+@Composable
+internal fun CompositionLocal<ComposeScene?>.requireCurrent(): ComposeScene {
+    return current ?: error("CompositionLocal LocalComposeScene not provided")
+}
+
+/**
+ * A virtual container that encapsulates Compose UI content. UI content can be constructed via
+ * [setContent] method and with any Composable that manipulates [LayoutNode] tree.
+ *
+ * After [ComposeScene] will no longer needed, you should call [close] method, so all resources
+ * and subscriptions will be properly closed. Otherwise, there can be a memory leak.
+ */
+@InternalComposeUiApi
+interface ComposeScene {
+    /**
+     * Density of the content which will be used to convert [Dp] units.
+     */
+    var density: Density
+
+    /**
+     * The layout direction of the content, provided to the composition via [LocalLayoutDirection].
+     */
+    var layoutDirection: LayoutDirection
+
+    /**
+     * Represents the scene size in scaled pixels ([Dp]).
+     * It's used to apply constrains to the content.
+     */
+    var size: IntSize?
+
+    /**
+     * The object that used to share "context" between multiple scenes on the screen.
+     * Also, it provides a way for platform interaction that required within a scene.
+     *
+     * @see ComposeSceneContext
+     */
+    val composeSceneContext: ComposeSceneContext
+
+    /**
+     * Top-level composition locals, which will be provided for the Composable content, which is set by [setContent].
+     *
+     * `null` if no composition locals should be provided.
+     */
+    var compositionLocalContext: CompositionLocalContext?
+
+    /**
+     * The last known position of pointer cursor position or `null` if cursor is not inside a scene.
+     */
+    val lastKnownPointerPosition: Offset?
```

Comment:
The goal of the scene is to be a facade to the platform, not to be facade to the internals of the scene.

The data flow is:
```
platform event -> scene -> lastKnownPointerPosition -> Composable functions
```

Let's remove it from the interface and provide some internal entity for `lastKnownPointerPosition` (internal LocalPointerPositions?)

Reply:
> internal entity for lastKnownPointerPosition (internal LocalPointerPositions?)

Or move it to PlatformContext, we don't need it to store it only in one scene, we need global state

Reply:
Changed to `BaseScene` cast in `rememberCursorPosition` (single usage)